scripts/add_brownfield.py

```diff
@@ -0,0 +1,183 @@
+#overriding the component_attrs. This can be done for
+#as many buses as you need with format busi for i = 2,3,4,5,....
+#See https://pypsa.org/doc/components.html#link-with-multiple-outputs-or-inputs
+
+override_component_attrs = pypsa.descriptors.Dict({k : v.copy() for k,v in pypsa.components.component_attrs.items()})
+override_component_attrs["Link"].loc["bus2"] = ["string",np.nan,np.nan,"2nd bus","Input (optional)"]
+override_component_attrs["Link"].loc["bus3"] = ["string",np.nan,np.nan,"3rd bus","Input (optional)"]
+override_component_attrs["Link"].loc["efficiency2"] = ["static or series","per unit",1.,"2nd bus efficiency","Input (optional)"]
+override_component_attrs["Link"].loc["efficiency3"] = ["static or series","per unit",1.,"3rd bus efficiency","Input (optional)"]
+override_component_attrs["Link"].loc["p2"] = ["series","MW",0.,"2nd bus output","Output"]
+override_component_attrs["Link"].loc["p3"] = ["series","MW",0.,"3rd bus output","Output"]
+
+
+def prepare_costs():
+
+    #set all asset costs and other parameters
+    #costs = pd.read_csv(snakemake.input.costs,index_col=list(range(3))).sort_index()
+    
+    costs = pd.read_csv(snakemake.input.costs,index_col=list(range(2))).sort_index()
+    
+    #correct units to MW and EUR
+    costs.loc[costs.unit.str.contains("/kW"),"value"]*=1e3
+    costs.loc[costs.unit.str.contains("USD"),"value"]*=snakemake.config['costs']['USD2013_to_EUR2013']
+
+    #cost_year = snakemake.config['costs']['year']
+    #costs = costs.loc[idx[:,cost_year,:],"value"].unstack(level=2).groupby(level="technology").sum(min_count=1)
+    costs = costs.loc[:, "value"].unstack(level=1).groupby("technology").sum()
+    costs = costs.fillna({"CO2 intensity" : 0,
+                          "FOM" : 0,
+                          "VOM" : 0,
+                          "discount rate" : snakemake.config['costs']['discountrate'],
+                          "efficiency" : 1,
+                          "fuel" : 0,
+                          "investment" : 0,
+                          "lifetime" : 25
+    })
+
+    costs["fixed"] = [(annuity(v["lifetime"],v["discount rate"])+v["FOM"]/100.)*v["investment"]*Nyears for i,v in costs.iterrows()]
+    return costs
+
+    
+def add_brownfield(n,n_p):
+    print("adding brownfield")
+    
+    previous_timestep=snakemake.config['scenario']['planning_horizons'][snakemake.config['scenario']['planning_horizons'].index(year)-1]
+    previous_timesteps=snakemake.config['scenario']['planning_horizons'][0:snakemake.config['scenario']['planning_horizons'].index(year)]
+
+    
+    #add generators from previous step     
+    n_p.mremove("Generator", [index for index in n_p.generators.index.to_list() if '<'+snakemake.config['scenario']['planning_horizons'][0] in index])
+    n_p.mremove("Generator", [index for index in n_p.generators.index.to_list() if 'ror' in index])
+        
+    n_p.generators.index=np.where(n_p.generators.index.str[-4:].isin(previous_timesteps)==False,
```

Comment:
I don't 100% understand what's going on here. So if it doesn't have a -year suffix in previous_timesteps, then we just assume it's from the previous year? What generators fall into this category? Don't we need to take account of the asset lifetime here? E.g. 20-year-old asset built in 2020 will not be there in 2050.

Reply:
I have improved the code and the commented lines that explain it:
1.	If it doesn’t have a -year suffix in previous_timesteps, we assume that it was optimized in the previous time_step
2.	We remove those generators whose installationYear + lifetime < year
3.	We rename capacities optimized in the previous timestep by adding the suffix eg. -2030 and add them to the network.

All the generators are included into this category (except ror which are removed since their capacities are exogenously fixed).

So far, I’m assuming a “default” lifetime (included in the config file). 
What would be the best way of recover technology names? so that we can read costs.at[technology, ‘lifetime’].

I was considering using the generators name or carrier but it is not direct. Maybe I should just wait until the attribute “lifetime” is added (I understood that this was the plan for the perfect foresight approach).